### PR TITLE
Apply

### PR DESCRIPTION
**Pull request by PatelSneh21**
I've added in the functionality to send emails into this branch. You can use the sendStatusEmail function to send an email using a given template. The next step is to create a template email for travel reimbursements and then to create a button that organizers can use to send all registrees an email to fill out the travel reimbursement form.

### CODE REVIEW

**Comment by nkalupahana** [comment]
Can you fix the conflicts + prettier stuff so we can do a review?

**Comment by PatelSneh21** [comment]
> Can you fix the conflicts + prettier stuff so we can do a review?

I think it should be okay now, there's a new dependency issue but I fixed the lockfile conflicts and prettier stuff. lmk if I need to do something for the dependency or if it will be resolved after the merge.

File: components/sendEmailForm.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { Form, Button, Select, Row, Col, message } from 'antd';
+import Text from 'antd/lib/typography/Text';
+import { useState } from 'react';
+
+const { Option } = Select;
+
+export interface ManageFormFields2 {
```
Comment (aadibajpai) [inline components/sendEmailForm.tsx:7]: should be called something better I don't think it's related to manage form

File: components/sendEmailForm.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	_id: string;
+	name: string;
+	email: string;
+	applicationStatus: string; //Enum for status needed (Submitted, Accepted, Declined)
+}
+
+export interface ManageFormProps2 {
```
Comment (aadibajpai) [inline components/sendEmailForm.tsx:14]: same here

File: components/sendEmailForm.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	applicationData: ManageFormFields2[];
+	onSubmit: (value: ManageFormFields2) => Promise<void>;
+}
+
+export default function ManageRoleForm(props: ManageFormProps2) {
```
Comment (aadibajpai) [inline components/sendEmailForm.tsx:19]: and here

File: pages/OrganizerDash.tsx [hunk 4]
```diff
@@ -106,6 +122,17 @@ export default function OrganizerDash() {
 		return (await res.json()) as { _id: string; name: string; email: string; userType: string }[];
 	});
 
+	const { data: statusData, error: statusError } = useSWR('/api/send-email', async url => {
```
Comment (aadibajpai) [inline pages/OrganizerDash.tsx:125]: where is this used and where is the component used?

File: components/sendEmailForm.tsx (hunk 1 above)
Comment (PatelSneh21) [reply components/sendEmailForm.tsx:7]: This was just there for testing purposes, will remove.

File: pages/api/send-email.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+export default async function handler(req: NextApiRequest, res: NextApiResponse<any>) {
+	const session = await getSession({ req });
+	if (session?.userType !== 'ORGANIZER') return res.status(403).send('Forbidden');
+	// TODO: add uniqueness validation try catch
+	await dbConnect();
+	switch (req.method) {
+		case 'GET':
+			const logins = await User.find({}).select('id name email applicationStatus');
+
+			return res.status(200).send(logins);
```
Comment (Ace314159) [inline pages/api/send-email.ts:18]: Why is this needed?

File: pages/api/send-email.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+			for (const user in userData) {
+				console.log(user, userData[user]);
+				statuses[userData[user] as statusIndex].push(user);
+				await User.updateOne();
+			}
+
+			// set their type to the one provided
+			for (const status in statuses) {
+				//await User.updateMany({ _id: { $in: statuses[status as statusIndex] } }, { applicationStatus: status });
```
Comment (Ace314159) [inline pages/api/send-email.ts:39]: Can you get the application status changing query to work as well? We should be doing both in one backend call. Then, I think you should change the file name to something like `update-application-status` or smth to indicate that.

File: pages/api/send-email.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	const session = await getSession({ req });
+	if (session?.userType !== 'ORGANIZER') return res.status(403).send('Forbidden');
+	// TODO: add uniqueness validation try catch
+	await dbConnect();
+	switch (req.method) {
+		case 'GET':
+			const logins = await User.find({}).select('id name email applicationStatus');
+
+			return res.status(200).send(logins);
+		case 'PATCH':
```
Comment (Ace314159) [inline pages/api/send-email.ts:19]: I feel like this should to be a post?
```suggestion
		case 'POST':
```

File: pages/api/send-travel-email.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+export default async function handler(req: NextApiRequest, res: NextApiResponse<any>) {
+	//const session = await getSession({ req });
+	//if (session?.userType !== 'ORGANIZER') return res.status(403).send('Forbidden');
+	// TODO: add uniqueness validation try catch
+	await dbConnect();
+	switch (req.method) {
+		case 'GET':
+			const logins = await User.find({}).select('id name email applicationStatus');
+
+			return res.status(200).send(logins);
```
Comment (Ace314159) [inline pages/api/send-travel-email.ts:19]: What is this for?

File: pages/api/send-travel-email.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import User from '../../models/user';
+import log from '../../middleware/log';
+import { ApplicationStatus, UserData } from '../../types/database';
+import { sendTravelEmail } from './emails/aws';
+import { UserSchema } from '../../models/user';
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse<any>) {
+	//const session = await getSession({ req });
+	//if (session?.userType !== 'ORGANIZER') return res.status(403).send('Forbidden');
+	// TODO: add uniqueness validation try catch
```
Comment (Ace314159) [inline pages/api/send-travel-email.ts:13]: We should add authentication. When is this going to be called, and who should be able to call it?

File: pages/api/logger.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Logger for server-side logging in JSON format
+ */
+
+import pino from 'pino';
```
Comment (Ace314159) [inline pages/api/logger.ts:5]: What is this?

File: pages/api/send-email.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+			const { applicationData: userData } = req.body;
+			if (!Object.keys(userData).length) return res.status(400).send('A user needs to be selected.');
```
Comment (Ace314159) [inline pages/api/send-email.ts:21]:
```suggestion
			if (!Object.keys(userData).length) return res.status(400).send('Users needs to be selected.');
```

File: pages/api/send-email.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+			const statuses = {
+				SUBMITTED: [] as (string | never)[],
+				ACCEPTED: [] as (string | never)[],
+				REJECTED: [] as (string | never)[],
+			};
+
+			type statusIndex = 'SUBMITTED' | 'ACCEPTED' | 'REJECTED';
+
+			for (const user in userData) {
+				console.log(user, userData[user]);
```
Comment (Ace314159) [inline pages/api/send-email.ts:32]: Get rid of your debugging logs

File: pages/api/send-email.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+				statuses[userData[user] as statusIndex].push(user);
+				await User.updateOne();
```
Comment (Ace314159) [inline pages/api/send-email.ts:34]: What is this doing? Nothing is passed in, so what is it updating?

File: pages/api/send-email.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+			}
+
+			// set their type to the one provided
+			for (const status in statuses) {
+				//await User.updateMany({ _id: { $in: statuses[status as statusIndex] } }, { applicationStatus: status });
+				//sendStatusEmail(user, status)
+				console.log(status);
```
Comment (Ace314159) [inline pages/api/send-email.ts:41]: Get rid of this log too

File: pages/api/send-email.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+			}
+
+			await log(session.userID, `Updated ${Object.keys(userData).length} user roles`);
+			return res.status(200).send(`Assigned status to ${userData.length} users`);
```
Comment (Ace314159) [inline pages/api/send-email.ts:45]:
```suggestion
			return res.status(200).send(`Assigned statuses to ${userData.length} users`);
```

**Review by Ace314159 — CHANGES_REQUESTED** [review]

File: pages/api/send-travel-email.ts (hunk 8 above)
Comment (PatelSneh21) [reply pages/api/send-travel-email.ts:19]: Needed it to get the user objects to pass in via postman, I can remove.
